docs/ORT_Use_Trtion_Kernel.md

```diff
@@ -0,0 +1,104 @@
+## Description
+
+In some scenarios, the Triton written kernels are more performant than CK or other handwritten kernels, so we implement a framework that enables onnxruntime to use these Triton written kernels.
+
+Here we use `softmax` op as an example to show how to integrate a Triton written kernel into onnxruntime CUDA/ROCm EP.
+
+### Write and compile Triton kernel
+
+We have implemented a softmax kernel using Triton at `onnxruntime/core/providers/rocm/math/softmax_triton.py`
+
+```python
+@triton.jit
+def softmax_kernel(
+    output_ptr, input_ptr, input_row_stride, output_row_stride, n_cols,
+    BLOCK_SIZE: tl.constexpr
+):
+    # softmax implementations
+    ...
+    ...
+```
+
+This is a very simple implementation. The `n_cols` parameter should be smaller than `BLOCK_SIZE`. And `BLOCK_SIZE` MUST be determined at compile time.
+
+In order to support different input shapes, we compile multiple kernels with different `BLOCK_SIZE`s.
+
+Each kernel with different `BLOCK_SIZE` generates different `num_warps` and shared memory usage, we call them `metadata`, and these metadata are needed when launching kernels in onnxruntime.
+
+We develop a script `tools/ci_build/compile_triton.py` to compile kernel and generate metadata for kernel launching.
+
+To generate metadata for softmax, we need to add description info and implement a `get_function_table` function in `softmax_triton.py`:
+
+```python
+# kernel dtype and BLOCK_SIZE to generate.
+dtypes = ['fp32', 'fp16']
+blocks = [1024, 2048, 4096, 8192, 16384]
+name_pattern = 'softmax_{}_{}'
+sig_pattern = '*{},*{},i32,i32,i32'
+group_pattern = 'softmax_{}'
+
+"""
+SHOULD implement a function that returns a metadata list with format:
+
+function_table = [
+    {
+        'name': xx,
+        'group': yy,
+        'func': func,
+        'sig': sig,
+        'kwargs': kwargs
+    }
+]
+
+The kwargs is a dict of {string: int} which is used for kernel constants. For example, BLOCK_SIZE of softmax.
+"""
+
+def get_function_table():
+    ...
+```
+
+When compiling onnxruntime with `--use_triton_kernel` flag, this softmax kernel will be compiled and combined into `libonnxruntime_providers_rocm.so` for ROCm or `libonnxruntime_providers_cuda.so` for CUDA.
```

Comment:
Are you sure that the kernel will be compiled into CUDA?

Reply:
Sorry I did not test it. This doc is renamed from an old one, I haven't tried on CUDA side.